### PR TITLE
[RPA-1666] External actions scheduler migration

### PR DESCRIPTION
**Pull request by Aeri4a**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

---

I agree to the terms of the RunBotics Contributor License Agreement.

### CODE REVIEW

File: runbotics/runbotics-scheduler/src/migrations/1725274222207-ActionFeatureKey.ts [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+                    ...authority.featureKeys,
+                    { name: FeatureKey.EXTERNAL_ACTION_READ }
+                ];
+                return authority;
+            });
+
+        await queryRunner.manager.save(Authority, updatedUser);
+    }
+
+    public async down(queryRunner: QueryRunner): Promise<void> {
```
Comment (star-in-the-jar) [inline runbotics/runbotics-scheduler/src/migrations/1725274222207-ActionFeatureKey.ts:36]: is this one tested?

File: runbotics/runbotics-scheduler/src/scheduler-database/action/action.service.ts [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+        id: string,
+        tenantId: string,
+        updateActionDto: UpdateActionDto,
+    ) {
+        const updatedAction = await this.actionRepository
+            .findOneByOrFail({ tenantId, id })
+            .then(action => ({
+                ...action,
+                ...updateActionDto,
+            })).catch(() => {
```
Comment (star-in-the-jar) [inline runbotics/runbotics-scheduler/src/scheduler-database/action/action.service.ts]: .catch to new line

File: runbotics/runbotics-scheduler/src/scheduler-database/action/dto/create-action.dto.ts [hunk 3]
```diff
@@ -0,0 +1,11 @@
+import { z } from 'zod';
+
+
+export const createActionSchema = z.object({
+    id: z.string().startsWith('external.').max(255),
+    script: z.string().startsWith('external.').max(255),
```
Comment (star-in-the-jar) [inline runbotics/runbotics-scheduler/src/scheduler-database/action/dto/create-action.dto.ts]: You can add validation error message if conditions are not met - so user can be informed that action has to start with 'external.'

Example: 
image

File: runbotics/runbotics-scheduler/src/scheduler-database/action/dto/update-action.dto.ts [hunk 4]
```diff
@@ -0,0 +1,9 @@
+import { z } from 'zod';
+
+export const updateActionSchema = z.object({
+    script: z.string().startsWith('external.').max(255),
```
Comment (star-in-the-jar) [inline runbotics/runbotics-scheduler/src/scheduler-database/action/dto/update-action.dto.ts]: same as above

**Review by star-in-the-jar — APPROVED** [review]

File: runbotics/runbotics-scheduler/src/scheduler-database/action/action.entity.ts [hunk 5]
```diff
@@ -0,0 +1,26 @@
+import { Column, Entity, JoinColumn, ManyToOne, PrimaryColumn, Unique } from 'typeorm';
+import { Tenant } from '../tenant/tenant.entity';
+
+
+@Entity({ name: 'action' })
+@Unique(['tenantId', 'id'])
+export class Action {
+    @PrimaryColumn({ type: 'varchar', length: 255 })
```
Comment (marcin-depcik) [inline runbotics/runbotics-scheduler/src/scheduler-database/action/action.entity.ts:9]: I think that `255` is a default length for type `varchar` in typeorm, isn't it?

File: runbotics/runbotics-scheduler/src/scheduler-database/action/action.entity.ts [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+    @Column({ type: 'varchar', length: 255 })
+    label: string;
+
+    @Column({ type: 'text' })
+    form: string;
+
+    @Column({ type: 'varchar', length: 255 })
+    script: string;
+
+    @Column({ name: 'tenant_id', type: 'uuid', default: 'b7f9092f-5973-c781-08db-4d6e48f78e98' })
```
Comment (marcin-depcik) [inline runbotics/runbotics-scheduler/src/scheduler-database/action/action.entity.ts]: I would assign it to some kind of const variable accessible for project as `DEFAULT_TENENT_ID` or something like that

File: runbotics/runbotics-scheduler/src/scheduler-database/action/action.entity.ts [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+    tenantId: string;
+
+    @ManyToOne(() => Tenant, tenant => tenant.id, { nullable: false })
```
Comment (marcin-depcik) [inline runbotics/runbotics-scheduler/src/scheduler-database/action/action.entity.ts]: `nullable: false` is also default value

File: runbotics/runbotics-scheduler/src/scheduler-database/action/action.entity.ts [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+import { Tenant } from '../tenant/tenant.entity';
+
+
+@Entity({ name: 'action' })
+@Unique(['tenantId', 'id'])
+export class Action {
+    @PrimaryColumn({ type: 'varchar', length: 255 })
+    id: string;
+
+    @Column({ type: 'varchar', length: 255 })
```
Comment (marcin-depcik) [inline runbotics/runbotics-scheduler/src/scheduler-database/action/action.entity.ts:12]: additionally, typeorm by default maps `string` as `type: 'varchar', length: 255`

File: runbotics/runbotics-scheduler/src/scheduler-database/action/dto/create-action.dto.ts (hunk 3 above)
Comment (marcin-depcik) [reply runbotics/runbotics-scheduler/src/scheduler-database/action/dto/create-action.dto.ts]: you can use `refine()` or message property 
image

**Review by marcin-depcik — CHANGES_REQUESTED** [review]

File: runbotics/runbotics-scheduler/src/migrations/1725274222207-ActionFeatureKey.ts (hunk 1 above)
Comment (Aeri4a) [reply runbotics/runbotics-scheduler/src/migrations/1725274222207-ActionFeatureKey.ts:36]: yes, it is

File: runbotics/runbotics-scheduler/src/scheduler-database/action/dto/create-action.dto.ts (hunk 3 above)
Comment (Aeri4a) [reply runbotics/runbotics-scheduler/src/scheduler-database/action/dto/create-action.dto.ts]: Good idea!

File: runbotics/runbotics-scheduler/src/scheduler-database/action/action.entity.ts (hunk 5 above)
Comment (Aeri4a) [reply runbotics/runbotics-scheduler/src/scheduler-database/action/action.entity.ts:9]: Unfortunately, you are not right. I met this case before, if we delete it then typeorm migration generator will force to drop the whole column, which is not okay :/

File: runbotics/runbotics-scheduler/src/scheduler-database/action/action.entity.ts (hunk 6 above)
Comment (Aeri4a) [reply runbotics/runbotics-scheduler/src/scheduler-database/action/action.entity.ts]: I will do it

File: runbotics/runbotics-scheduler/src/scheduler-database/action/action.entity.ts (hunk 8 above)
Comment (Aeri4a) [reply runbotics/runbotics-scheduler/src/scheduler-database/action/action.entity.ts:12]: That is not true. Probably the length is not set as fixed and that's why generator can see the difference. Those lengths are probably leftovers after Java.

**Review by marcin-depcik — APPROVED** [review]

**Pull request merged**